enterprise/glassfish.common/src/org/netbeans/modules/glassfish/common/StartTask.java

```diff
@@ -757,6 +761,20 @@ private void appendStartupExtenderParams(List<String> optList) {
         }
     }
 
+    /**
+     * Appends options in order to open modules for java 9.0+
+     * @param optList 
+     */
+    private void appendModuleOpensParams(List<String> optList) {
+        SpecificationVersion currentInstanceJavaVersion = JavaUtils.serverInstancePlatform(instance).getSpecification().getVersion();
+        SpecificationVersion firstJavaVersionWithModules = new SpecificationVersion("9");
+
+        if(currentInstanceJavaVersion.compareTo(firstJavaVersionWithModules)>=0){        
+            optList.add("--add-opens java.base/java.lang=ALL-UNNAMED");
+            optList.add("--add-opens java.naming/javax.naming.spi=ALL-UNNAMED");
```

Comment:
Why is this necessary? I downloaded Glassfish 6.1 from github and was able to start it standalone with JDK 11. This lead me to have a look at the domain.xml and I found this:

```xml
        <jvm-options>--add-opens=jdk.management/com.sun.management.internal=ALL-UNNAMED</jvm-options>
        <jvm-options>--add-opens=java.base/sun.net.www.protocol.jrt=ALL-UNNAMED</jvm-options>
        <jvm-options>--add-opens=java.base/java.lang=ALL-UNNAMED</jvm-options>
        <jvm-options>--add-opens=java.base/java.util=ALL-UNNAMED</jvm-options>
        <jvm-options>--add-opens=java.rmi/sun.rmi.transport=ALL-UNNAMED</jvm-options>
```

The jvm-options are picked up by:

https://github.com/apache/netbeans/blob/master/enterprise/glassfish.tooling/src/org/netbeans/modules/glassfish/tooling/server/ServerTasks.java#L162
and
https://github.com/apache/netbeans/blob/master/enterprise/glassfish.tooling/src/org/netbeans/modules/glassfish/tooling/server/parser/JvmConfigReader.java#L187-L189
https://github.com/apache/netbeans/blob/master/enterprise/glassfish.tooling/src/org/netbeans/modules/glassfish/tooling/server/parser/JvmConfigReader.java#L134-L139

As the server start standalone, I don't see why netbeans needs the addtional opens. Can you shed some light on this?

Reply:
Initially I saw the same thing, and tried running without this part. I am using Liberica Jdk16, and downloaded and installed a fresh GF 6.1 directly from Netbeans, without running Derby. I ran it and got explicit errors that said that opening was needed. I did not investigate further for lack of time.

Reply:
The issue is here:

https://github.com/apache/netbeans/blob/19be548f0bc53ecac53122e478a37e27ce344118/enterprise/glassfish.tooling/src/org/netbeans/modules/glassfish/tooling/server/ServerTasks.java#L304-L368

This takes all parameters, splits them into key-value parts, stuffs them into a map, modifies them and serializes them into the command line. The problem is, that putting the options into a map breaks all options, that can be set multiple times., as "add-opens" can. Only one value survives this procedure. It is just luck, that that trick works at all. To me that function needs to be fixed (a multimap?).